server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -1223,4 +1211,37 @@ static class PipelineHolder {
         }
     }
 
+    private static IndexMetadata findIndexMetadata(
+        DocWriteRequest<?> originalRequest,
+        IndexRequest indexRequest,
+        Metadata clusterMetadata,
+        long epochMillis
+    ) {
+        IndexMetadata indexMetadata = null;
+        // start to look for default or final pipelines via settings found in the cluster metadata
+        if (originalRequest != null) {
+            indexMetadata = clusterMetadata.indices()
+                .get(IndexNameExpressionResolver.resolveDateMathExpression(originalRequest.index(), epochMillis));
+        }
+        // check the alias for the index request (this is how normal index requests are modeled)
+        if (indexMetadata == null && indexRequest.index() != null) {
+            IndexAbstraction indexAbstraction = clusterMetadata.getIndicesLookup().get(indexRequest.index());
+            if (indexAbstraction != null && indexAbstraction.getWriteIndex() != null) {
+                indexMetadata = clusterMetadata.index(indexAbstraction.getWriteIndex());
+            }
+        }
+        // check the alias for the action request (this is how upserts are modeled)
+        if (indexMetadata == null && originalRequest != null && originalRequest.index() != null) {
+            IndexAbstraction indexAbstraction = clusterMetadata.getIndicesLookup().get(originalRequest.index());
+            if (indexAbstraction != null && indexAbstraction.getWriteIndex() != null) {
+                indexMetadata = clusterMetadata.index(indexAbstraction.getWriteIndex());
+            }
+        }
+        return indexMetadata;
+    }
+
+    public static boolean indexRequestHasPipeline(IndexRequest indexRequest) {
```

Comment:
I'm not sure, but perhaps this belongs on `IndexRequest` itself? Given that `IndexRequest` already has `isPipelineResolved`, maybe this makes sense there -- note that I'm not saying that it definitely does.

Reply:
Yes, I concur with you on this, but I'd do that in a different PR because I'd like to group several pipeline-related concepts in their own classes/records and clarify the domain. Let's discuss this offline later.